src/engine/song.cpp

```diff
@@ -102,6 +102,211 @@ bool DivSubSong::walk(int& loopOrder, int& loopRow, int& loopEnd, int chans, int
   return false;
 }
 
+double calcRowLenInSeconds(const DivGroovePattern& speeds, float hz, int vN, int vD, int timeBaseFromSong) 
+{
+    double hl=1; //count for 1 row
+  if (hl<=0.0) hl=4.0;
+  double timeBase=timeBaseFromSong+1;
+  double speedSum=0;
+  for (int i=0; i<MIN(16,speeds.len); i++) {
+    speedSum+=speeds.val[i];
+  }
+  speedSum/=MAX(1,speeds.len);
+  if (timeBase<1.0) timeBase=1.0;
+  if (speedSum<1.0) speedSum=1.0;
+  if (vD<1) vD=1;
+  //return (60.0 * hz / (timeBase * hl * speedSum)) * (double)vN / (double)vD;
+  return 1.0 / ((60.0*hz/(timeBase*hl*speedSum))*(double)vN/(double)vD / 60.0);
+}
+
+void DivSubSong::findLength(int loopOrder, int loopRow, double fadeoutLen, int& rowsForFadeout, bool& hasFFxx, std::vector<int>& orders_vec, std::vector<DivGroovePattern>& grooves, int& length, int chans, int jumpTreatment, int ignoreJumpAtEnd, int firstPat)
+{
+  length = 0;
+  hasFFxx = false;
+  rowsForFadeout = 0;
+
+  float secondsPerThisRow = 0.0f;
+
+  DivGroovePattern curSpeeds = speeds; //simulate that we are playing the song, track all speed/BPM/tempo/engine rate changes
+  short curVirtualTempoN = virtualTempoN;
+  short curVirtualTempoD = virtualTempoD;
+  float curHz = hz;
+  double curDivider = (double)timeBase;
+
+  double curLen = 0.0; //how many seconds passed since the start of song
```

Comment:
Hmmmm........... Using doubles and floats for time. Not sure whether it's a good idea as floating-point numbers lose precision.

Reply:
It's your BPM calc and engine code but adapted to calculate how many rows starting from song loop point will fadeout take. I understand that floating point loses precision, but here 1 row usually takes less than 1 second ao I decided to use floating point. By the way, this code averages groove's speed for each row so it may introduce some error to time calc if e.g. groove is 4 steps long and groove works over the course of e.g. 13 ticks before switching to constant speed again. But I think it's negligible most of the time.

Reply:
Fadeout is in double precision floating point variable anyway